cookbooks/bcpc/recipes/serial_console.rb

```diff
@@ -0,0 +1,72 @@
+require 'mixlib/shellout'
+
+# declare a placeholder template
+template "ttySx_svc_conf" do
+  source "ttySx.erb"
+  action :nothing
+end
+
+# declare a placeholder service
+service "ttySx_svc" do
+  action :nothing
+  provider Chef::Provider::Service::Upstart
+end
+
+ruby_block "discover_serial" do
+  block do
+    def numeric?(data)
+      Float(data) != nil rescue false
+    end
+    # we are not interested in USB or any other serial at the moment
+    # Sample file format where the first column
+    # number x corresponds to ttySx
+    # serinfo:1.0 driver revision:
+    # 0: uart:16550A port:000003F8 irq:4 tx:165 rx:0 CTS|DSR|CD
+    # 1: uart:unknown port:000002F8 irq:3
+    # 2: uart:unknown port:000003E8 irq:4
+    serial_dev_type = "ttyS"
+    serial_driver_info = "/proc/tty/driver/serial"
+    File.open(serial_driver_info, "r") do |serials|
+      serials.each_line do |line|
+        if line.start_with? "serinfo:"
+          next
+        end
+        line_parts = line.split()
+        uart = line_parts[1]
+        if not uart.include? "uart:unknown"
+          port_assignment = line_parts[0]
+          port_number = port_assignment.chop
+          # get device baud we expect
+          # root@bcpc-vm2:/proc/tty/driver# stty -F /dev/ttyS0 speed
+          # 115200
+          device_baud_cmd_str = ["/bin/stty", "-F", "/dev/#{serial_dev_type}#{port_number}", "speed"]
+          device_baud_cmd = Mixlib::ShellOut.new(device_baud_cmd_str, :timeout=>30)
+          device_baud_cmd.run_command
+          device_baud = device_baud_cmd.stdout.chomp
+          if not numeric?(device_baud)
+            next
+          end
+          device_name = "#{serial_dev_type}#{port_number}"
+          Chef::Log.info("Serial device found: #{device_name}:#{device_baud}")
+          Chef::Log.info("Running template for: #{device_name}:#{device_baud}")
+          # load the template resource and modify
+          tmpl = run_context.resource_collection.find("ttySx_svc_conf]")
+          tmpl.path("/etc/init/#{device_name}.conf")
+          tmpl.source("ttySx.erb")
+          tmpl.variables(:serial_params => {
+            "serial_device" => device_name,
+            "device_baud" => device_baud })
+          tmpl.owner("root")
+          tmpl.group("root")
+          tmpl.mode(00644)
+          tmpl.run_action(:create)
+          # load the service resource and create
+          Chef::Log.info("Creating service for: #{device_name}:#{device_baud}")
+          svc = run_context.resource_collection.find("service[ttySx_svc]")
+          svc.service_name(device_name)
+          svc.run_action(:start)
```

Comment:
There is no need, the job already defines run levels.  Unlike SysV init, in upstart  there are no run level specific directories to create symlinks